juju/bundle.py

```diff
@@ -156,7 +177,100 @@ async def fetch_plan(self, entity_id):
         if self.plan.errors:
             raise JujuError(self.plan.errors)
 
+    async def _download_bundle(self, charm_url, origin):
+        if self.charms_facade is not None:
```

Comment:
This would be cleaner if you did:

```
if self.charms_facade is None:
    raise ...

resp = await self.charms_facade ...
```

With Python, I like to see code that uses every opportunity to avoid indenting. And also it's nice to have the Exceptions thrown as close as possible, visually, to the check that causes it.